File: boj_codes/C++17/1927.cpp
// https://www.acmicpc.net/problem/1927
// First Written : 20210101
// Last Modified : 20210101

#include <cstdio>
#include <vector>

using namespace std;

class Heap{
private:
    int* heaps;
    int size;

public:
    Heap(int N){
        heaps = new int[N]();
        size = 0;
    }

    ~Heap(){
        delete [] heaps;
    }

    void push(int val){
        heaps[++size] = val;
        int idx = size;
        
        while(idx != 1){
            if(heaps[idx] < heaps[get_parent(idx)]){
                swap(idx, get_parent(idx));
                idx = get_parent(idx);
            }

            else{
                break;
            }

        }

    }

    int pop(){
        if(size == 0)
            return 0;
        
        int val = heaps[1];
        heaps[1] = heaps[size--];

        int idx = 1;
        int idx_c = get_Schild(idx);

        while(idx_c > 0){
            if(heaps[idx_c] < heaps[idx]){
                swap(idx, idx_c);
                idx = idx_c;
                idx_c = get_Schild(idx);
            }
            else{
                break;
            }
        }

        return val;
    }

    static inline int get_parent(int idx){
        return idx / 2;
    }

    static inline int get_Lchild(int idx){
        return idx * 2;
    }

    static inline int get_Rchild(int idx){
        return (idx * 2) + 1;
    }

    int get_Schild(int idx){
        int L_c = get_Lchild(idx);
        int R_c = get_Rchild(idx);

        if(L_c > size){
            return -1;
        }
        else if(R_c > size){
            return L_c;
        }

        else{
            if(heaps[L_c] <= heaps[R_c])
                return L_c;
            
            else
                return R_c;
            
        }

    }

    void swap(int i, int j){
        int temp = heaps[i];
        heaps[i] = heaps[j];
        heaps[j] = temp;
    }



};

int main(){
    int N, val;
    scanf("%d", &N);

    Heap heap = Heap(N);

    while(N--){
        scanf("%d", &val);
        if(val == 0){
            printf("%d\n", heap.pop());
        }

        else
            heap.push(val);

    }

    return 0;
}

File: boj_codes/Python3/1927.py
# https://www.acmicpc.net/problem/1927
# 20201016

import sys

class m_Heap:
    def __init__(self):
        self.heap = [0]
        self.len = 0

    def insert(self, new_val):
        self.heap.append(new_val)
        self.len += 1

        self.heapify(index= self.len, mode = 'i')

    def delete(self):
        if self.len == 0:
            return 0

        val = self.heap[1]
        if self.len == 1:
            self.heap.pop()
        else:
            self.heap[1] = self.heap.pop()

        self.len -= 1
        self.heapify(index=1, mode = 'd')

        return val

    def get_parent(self, index):
        return index // 2

    def get_Lchild(self, index):
        return index * 2

    def get_Rchild(self, index):
        return index * 2 + 1

    def get_Schild(self, index):
        L_child = self.get_Lchild(index)
        R_child = self.get_Rchild(index)

        if L_child > self.len:
            return -1

        elif R_child > self.len:
            return L_child

        else:
            if self.heap[L_child] <= self.heap[R_child]:
                return L_child
            else:
                return R_child

    def swap_index(self, i, j):
        temp = self.heap[i]
        self.heap[i] = self.heap[j]
        self.heap[j] = temp

    def heapify(self, index, mode):
        if mode == 'i':
            if index == 1:
                return

            index_p = self.get_parent(index)
            if self.heap[index] < self.heap[index_p]:
                self.swap_index(index, index_p)
                return self.heapify(index_p, mode)

            else:
                return

        if mode == 'd':
            index_c = self.get_Schild(index)
            if index_c < 0:
                return

            elif self.heap[index_c] < self.heap[index]:
                self.swap_index(index_c, index)
                self.heapify(index_c, mode)

N = int(sys.stdin.readline().rstrip())
heap = m_Heap()

for _ in range(N):
    val = int(sys.stdin.readline().rstrip())

    if val == 0:
        print(heap.delete())

    else:
        heap.insert(val)

